### DISCUSSION TITLE
How to eplore defined elements in a Rust program via Rustic?

### DISCUSSION THREAD

**Discussion by parvizfarnia** [brotzeit/rustic#500] [Q&A]
Hi,

With IDE's such as Netbeans, Eclipse, VSCode, etc. one has the possibility of having an explorer as a sidebar which includes the entire project folder with all source files but also everything else: all defined variables, classes, methods, functions, etc.

I use Emacs `Dired` for navigating between files in my project which works fine but for the rest, that is, accessing rapidly different element defined in my Rust programs: Structs, their members, Enums, functions, Traits, etc. so far, I haven't found a user-friendly navigation tool (sidebar).

I checked [Rustic's documentation](https://github.com/brotzeit/rustic), the closest thing that I found was using `xref (M-.)` by selecting a given word in the code. But even that, doesn't seem to work always correctly and does not necessarily take me to the exact location where the specific element is defined in my code.

So here is my question: Is it me who doesn't know how to use `rustic` efficiently or it's simply the fact that such functionnalities in terms of code navigation as I explained above are simply missing in `rustic` (at least in its current version)?

Thansk in advance

**Comment by psibi** [comment]
Probably you are looking for lsp-treemacs: https://github.com/emacs-lsp/lsp-treemacs#lsp-treemacs-symbols ?

**Reply by parvizfarnia** [reply]
Seems interesting. There are elements which are present but many (structures and enums) are absent. So if I understand correctly `rustic` itself doesn't provide this navigation functionality that I explained above, right?

**Reply by psibi** [reply]
Yeah, you're right.

> But even that, doesn't seem to work always correctly and does not necessarily take me to the exact location where the specific element is defined in my code.

I use rustic along with `lsp-mode` (and rust-analyzer) and the lsp's provided functionality for goto definition, finding references and finding type definitions have worked at exact locations for me.
